### PR TITLE
[fx] Add a pass to fetch attributes of nn.Module to fx.node

### PR DESCRIPTION
**Pull request by 842974287**
Summary: Fetch the parameters that are needed for lowering from nn.Module to fx.node for leaf_modules.
A `module_fetch_book` maintains the names of attribute that we want to fetch for a specific nn.Module. `_version` is used to ensure that the fetch book is aligned with the module.

Test Plan: A test `test_fetch` is added to test_fx_experimental.py.

Differential Revision: D24957142

### CODE REVIEW

File: torch/fx/experimental/param_fetch.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
+from torch.fx.graph_module import GraphModule
+from typing import Any, Dict, List, Tuple
+import torch.nn as nn
+
+
+"""This dict maps the nn.Module class name to the attribute name list that we want to fetch for lowering.
+The first integer in the tuple is the version number of the nn.Module class when we create the parameter list.
+If there's a version mismatch then it means the parameter names in the book might be mismatched with nn.Module.
+"""
+module_fetch_book: Dict[str, Tuple[int, List[str]]] = {
```
Comment (jamesr66a) [inline torch/fx/experimental/param_fetch.py:10]: You should key on the `Type`s themselves, not their string `__name__`:

```
from typing import Type

<...>

module_fetch_book: Dict[Type, Tuple[int, List[str]]] = {
    torch.nn.Linear: (1, ["weight", "bias"]),
    torch.nn.Conv2d: (1, ["weight", "bias", "kernel_size", "stride", "padding", "dilation", "groups", "padding_mode"]),
    torch.nn.BatchNorm2d: (2, ["weight", "bias", "running_mean", "running_var"]),
}
```

Using `__name__` isn't safe, because it's only the last atom of the qualified name. Alternatively we could match on `torch.typename(m)`, but it's probably better to not project into string space at all and just match directly on types

File: torch/fx/experimental/param_fetch.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    "Conv2d": (1, ["weight", "bias", "kernel_size", "stride", "padding", "dilation", "groups", "padding_mode"]),
+    "BatchNorm2d": (2, ["weight", "bias", "running_mean", "running_var"]),
+}
+
+def fetch_module(mod: nn.Module) -> Dict[str, Any]:
+    """If `mod` is in `module_fetch_book`, fetch the mod's attributes that in the `module_fetch_book`
+    after checking module's version is compatible with the `module_fetch_book`.
+    """
+    parameters: Dict[str, Any] = {}
+    parameters["name"] = type(mod).__name__
```
Comment (jamesr66a) [inline torch/fx/experimental/param_fetch.py:21]:
```
    parameters["name"] = torch.typename(mod)
```

To get the fully-qualified name

File: torch/fx/experimental/param_fetch.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+    if type(mod).__name__ in module_fetch_book:
```
Comment (jamesr66a) [inline torch/fx/experimental/param_fetch.py:23]: With the change above, this will become

```
    if type(mod) in module_fetch_book:
        _version, param_to_fetch = module_fetch_book[type(mod)]

```

File: torch/fx/experimental/param_fetch.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        _version, param_to_fetch = module_fetch_book[type(mod).__name__]
```
Comment (jamesr66a) [inline torch/fx/experimental/param_fetch.py:24]: Convention is leading underscore indicates a private name, but since this is a local variable that doesn't really make sense, so let's just name this `version`

File: torch/fx/experimental/param_fetch.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        if _version != mod._version:
+            print(f"Fetcher version {_version} try to fetch {type(mod).__name__} version {mod._version}")
```
Comment (jamesr66a) [inline torch/fx/experimental/param_fetch.py:26]: Is this an error condition? If so,

```
            raise RuntimeError(f"Fetcher version {_version} try to fetch {type(mod).__name__} version {mod._version}")
```

If not, use `warnings.warn`.

File: torch/fx/experimental/param_fetch.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    "Linear": (1, ["weight", "bias"]),
+    "Conv2d": (1, ["weight", "bias", "kernel_size", "stride", "padding", "dilation", "groups", "padding_mode"]),
+    "BatchNorm2d": (2, ["weight", "bias", "running_mean", "running_var"]),
+}
+
+def fetch_module(mod: nn.Module) -> Dict[str, Any]:
+    """If `mod` is in `module_fetch_book`, fetch the mod's attributes that in the `module_fetch_book`
+    after checking module's version is compatible with the `module_fetch_book`.
+    """
+    parameters: Dict[str, Any] = {}
```
Comment (jamesr66a) [inline torch/fx/experimental/param_fetch.py:20]: The term "parameter" has a very specific meaning in PyTorch (trainable tensors), so please call this something else like `attrs`

File: torch/fx/experimental/param_fetch.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+The first integer in the tuple is the version number of the nn.Module class when we create the parameter list.
+If there's a version mismatch then it means the parameter names in the book might be mismatched with nn.Module.
+"""
+module_fetch_book: Dict[str, Tuple[int, List[str]]] = {
+    "Linear": (1, ["weight", "bias"]),
+    "Conv2d": (1, ["weight", "bias", "kernel_size", "stride", "padding", "dilation", "groups", "padding_mode"]),
+    "BatchNorm2d": (2, ["weight", "bias", "running_mean", "running_var"]),
+}
+
+def fetch_module(mod: nn.Module) -> Dict[str, Any]:
```
Comment (jamesr66a) [inline torch/fx/experimental/param_fetch.py:16]: "Fetch module" implies you're going to return a `Module` instance, but you're not. This is probably better named something like `extract_attrs_for_lowering`

File: torch/fx/experimental/param_fetch.py [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    if type(mod).__name__ in module_fetch_book:
+        _version, param_to_fetch = module_fetch_book[type(mod).__name__]
+        if _version != mod._version:
+            print(f"Fetcher version {_version} try to fetch {type(mod).__name__} version {mod._version}")
+        for attr in param_to_fetch:
+            parameters[attr] = getattr(mod, attr)
+
+    return parameters
+
+def fetch(fx_module: GraphModule):
```
Comment (jamesr66a) [inline torch/fx/experimental/param_fetch.py:32]: Similarly, this could be named better. Something like `lift_lowering_parameters_to_nodes`

File: torch/fx/experimental/param_fetch.py (hunk 5 above)
Comment (jamesr66a) [reply torch/fx/experimental/param_fetch.py:26]: We should also be checking that the `Module` we're processing has `version` at *or below* the one we support. Right now, with the 3 modules supported (Linear, Conv2d, and BatchNorm2d) it doesn't matter because the attributes of interest are all the same, but in the future we may need `module_fetch_book` to also contain functions that normalize the retrieved attributes across older versions

**Review by jamesr66a — COMMENTED** [review]
Cool!

File: torch/fx/experimental/param_fetch.py (hunk 5 above)
Comment (842974287) [reply torch/fx/experimental/param_fetch.py:26]: >contain functions that normalize the retrieved attributes across older versions

Sorry I'm a little bit confused. Do you mean when we processing `Module` with older versions, the function would come out and match the names of attribute in current version to the older version?

Comment (jamesr66a) [reply torch/fx/experimental/param_fetch.py]: Yes, that's correct

File: test/test_fx_experimental.py [hunk 9]
```diff
@@ -550,6 +551,28 @@ def foo(x):
...
+                a += a
+                return self.bn(a)
+
+        mod = TestModule()
+        traced = symbolic_trace(mod)
+        lift_lowering_attrs_to_nodes(traced)
+        for node in traced.graph.nodes:
+            if node.op == "call_module":
+                assert hasattr(node, "attrs_for_lowering")
+                _, para_list = module_fetch_book[locate(node.attrs_for_lowering["name"])]
```
Comment (jamesr66a) [inline test/test_fx_experimental.py:572]: As part of good testing practice, you shouldn't use the code you're trying to test as a reference in the test. I suggest just having a list of expected attributes for the Conv and BatchNorm node and assert that they are present manually.

File: torch/fx/node.py [hunk 10]
```diff
@@ -53,6 +53,9 @@ def __init__(self, graph: 'Graph', name: str, op: str, target: Target,
         # does not produce a value, it's more of a notation. Thus, this value
         # describes the type of args[0] in the `return` node.
         self.type : Optional[Any] = type
+        # Attrbutes that are needed to lowering the fx node.
+        # This dict is filled by param_fetch.
+        self.attrs_for_lowering : Dict[str, Any] = {}
```
Comment (jamesr66a) [inline torch/fx/node.py:58]: Please don't add this code to FX core, it is not generally applicable. You can either monkey-patch this attribute, or create a subclass of `Node` to inject into your `Graph`

File: torch/fx/experimental/param_fetch.py (hunk 5 above)
Comment (842974287) [reply torch/fx/experimental/param_fetch.py]: I've added a function to each entry. Right now for these three modules the function doesn't actually do anything.

File: torch/fx/node.py (hunk 10 above)
Comment (842974287) [reply torch/fx/node.py:58]: Ah I see, I'll try monkey-patch first (hopefully no mypy error

File: test/test_fx_experimental.py [hunk 11]
```diff
@@ -595,6 +596,39 @@ def foo(x):
...
+                return self.bn(a)
+
+        mod = TestModule()
+        traced = symbolic_trace(mod)
+        lift_lowering_attrs_to_nodes(traced)
+
+        for node in traced.graph.nodes:
+            if node.op == "call_module":
+                assert hasattr(node, "attrs_for_lowering")
+                para_list = attrs_for_lowering[locate(node.attrs_for_lowering["name"])]
```
Comment (jamesr66a) [inline test/test_fx_experimental.py:625]: How about we just use string keys in `attrs_for_lowering` in this test instead of using `locate`?

Comment (842974287) [reply test/test_fx_experimental.py]: Sounds good!

File: torch/fx/experimental/graph_manipulation.py [hunk 12]
```diff
@@ -122,19 +123,17 @@ def serialize_weight(tensor: torch.Tensor) -> Dict:
...
-        weights_metadata[f"{name_prefix}.{name}"] = serialize_weight(buffer)
-        weights[f"{name_prefix}.{name}"] = buffer
-    for name, parameter in mod.named_parameters():
-        weights_metadata[f"{name_prefix}.{name}"] = serialize_weight(parameter)
-        weights[f"{name_prefix}.{name}"] = parameter
-    if isinstance(mod.__constants__, List):
-        for constant in mod.__constants__:
-            parameters[constant] = str(getattr(mod, constant))
+
+    for p_name, p_value in node.attrs_for_lowering.items():  # type: ignore
```
Comment (jamesr66a) [inline torch/fx/experimental/graph_manipulation.py:130]: Why is there a `# type: ignore` here?

File: torch/fx/experimental/param_fetch.py [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+    """If `mod` is in `module_fetch_book`, fetch the mod's attributes that in the `module_fetch_book`
+    after checking module's version is compatible with the `module_fetch_book`.
+    """
+    attrs_for_lowering: Dict[str, Any] = {}
+    attrs_for_lowering["name"] = torch.typename(mod)
+
+    if type(mod) in module_fetch_book:
+        version, param_to_fetch, matching_method = module_fetch_book[type(mod)]
+        if version < mod._version:
+            raise RuntimeError(f"Fetcher version {version} try to fetch {torch.typename(mod)} version {mod._version}")
```
Comment (jamesr66a) [inline torch/fx/experimental/param_fetch.py:34]: This error message should also include instructions for what to do to proceed. In this case, likely the user should report a bug to the accelerator team

**Review by jamesr66a — COMMENTED** [review]
LGTM! @gcatron @jfix71 do you want to take a look?

File: torch/fx/experimental/param_fetch.py [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+def extract_attrs_for_lowering(mod: nn.Module) -> Dict[str, Any]:
+    """If `mod` is in `module_fetch_book`, fetch the mod's attributes that in the `module_fetch_book`
+    after checking module's version is compatible with the `module_fetch_book`.
+    """
+    attrs_for_lowering: Dict[str, Any] = {}
+    attrs_for_lowering["name"] = torch.typename(mod)
+
+    if type(mod) in module_fetch_book:
```
Comment (jfix71) [inline torch/fx/experimental/param_fetch.py:31]: Should we raise an error if we don't handle the type?

**Review by jfix71 — COMMENTED** [review]
Cool! Mostly LGTM, one comment

File: torch/fx/experimental/graph_manipulation.py (hunk 12 above)
Comment (842974287) [reply torch/fx/experimental/graph_manipulation.py:130]: Oh this is because I'm doing monkey-patch to add `attrs_for_lowering` which would cause the mypy test fail with error `node doesn't has attribute attrs_for_lowering`.

File: torch/fx/experimental/param_fetch.py (hunk 13 above)
Comment (842974287) [reply torch/fx/experimental/param_fetch.py:34]: Sounds good!

File: torch/fx/experimental/param_fetch.py (hunk 14 above)
Comment (842974287) [reply torch/fx/experimental/param_fetch.py:31]: Sounds good!

File: torch/fx/experimental/param_fetch.py [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+        version, param_to_fetch, matching_method = module_fetch_book[type(mod)]
+        if version < mod._version:
+            raise RuntimeError(f"Fetcher version {version} try to fetch {torch.typename(mod)} version {mod._version}, "
+                               "please upgrade the module_fetch_book or report a bug to AIACC team.")
```
Comment (jackm321) [inline torch/fx/experimental/param_fetch.py:40]: folks in OSS won't know who AIACC team is. Should we just ask for a gh issue or something?

Comment (842974287) [reply torch/fx/experimental/param_fetch.py:40]: Sounds good! I added another option as "open an issue and @842974287".

File: torch/fx/experimental/param_fetch.py [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+            raise RuntimeError(f"Fetcher version {version} try to fetch {torch.typename(mod)} version {mod._version}, "
+                               "please upgrade the module_fetch_book, open an issue and @842974287"
+                               "or report a bug to AIACC team directly.")
+        for attr in param_to_fetch:
+            attrs_for_lowering[attr] = getattr(mod, matching_method(attr, mod._version))
+    else:
+        raise RuntimeError(f"{torch.typename(mod)} is not in the module_fetch_book yet, "
+                           "please add it to the module_fetch_book, open an issue and @842974287"
+                           "or report a bug to AIACC team directly.")
+    return attrs_for_lowering
```
Comment (jfix71) [inline torch/fx/experimental/param_fetch.py:48]: nit: You didn't add a space between `@842974287` and the next line `or`, so it'll appear as `@842974287or` 🙂

**Review by jfix71 — APPROVED** [review]
🚀

File: torch/fx/experimental/param_fetch.py (hunk 16 above)
Comment (842974287) [reply torch/fx/experimental/param_fetch.py]: Nice catch!
